File: dev/20190218/k2works/fizz_buzz.py
import unittest


class FizzBuzzTest(unittest.TestCase):
    def test_1から100までの数をプリントする_ただし3で割り切れる場合はFizz_5で割り切れる場合はBuzz_両方で割り切れる場合はFizzBuzzをプリントする(self):
        fizz_buzz_list = FizzBuzz.generate_list()
        self.assertEqual(1, fizz_buzz_list[1])
        self.assertEqual('Fizz', fizz_buzz_list[3])
        self.assertEqual('Buzz', fizz_buzz_list[5])
        self.assertEqual('FizzBuzz', fizz_buzz_list[15])
        self.assertEqual('Buzz', fizz_buzz_list[-1])

    def test_3で割り切れる場合はFizzを返す(self):
        self.assertEqual('Fizz', FizzBuzz.generate(3))
        self.assertEqual(1, FizzBuzz.generate(1))

    def test_5で割り切れる場合はBuzzを返す(self):
        self.assertEqual('Buzz', FizzBuzz.generate(5))

    def test_3と5で割り切れる場合はFizzBuzzを返す(self):
        self.assertEqual('FizzBuzz', FizzBuzz.generate(15))

    def test_数字以外の場合は例外を発生させる(self):
        with self.assertRaises(ValueError):
            FizzBuzz.generate('a')


MAX_RANGE = 101
BUZZ = 'Buzz'
FIZZ = 'Fizz'
FIZZ_BUZZ = 'FizzBuzz'


class FizzBuzz:
    @staticmethod
    def print():
        try:
            [print(v) for v in FizzBuzz.generate_list()]
        except ValueError as e:
            print(e)

    @staticmethod
    def generate_list():
        return list(map(FizzBuzz.generate, range(MAX_RANGE)))

    @staticmethod
    def generate(number):
        if isinstance(number, str):
            raise ValueError

        if (FizzBuzz.is_fizz(number)) and (FizzBuzz.is_buzz(number)):
            return FIZZ_BUZZ
        if FizzBuzz.is_fizz(number):
            return FIZZ
        if FizzBuzz.is_buzz(number):
            return BUZZ
        return number

    @staticmethod
    def is_buzz(number):
        return number % 5 == 0

    @staticmethod
    def is_fizz(number):
        return number % 3 == 0


if __name__ == "__main__":
    unittest.main()

File: dev/20190218/k2works/fizz_buzz.rb
# frozen_string_literal: true

require 'test/unit'

class TestFizzBuzz < Test::Unit::TestCase
  test '1から100までの数をプリントする,ただし3で割り切れる場合はFizz,5で割り切れる場合はBuzz,両方で割り切れる場合はFizzBuzzをプリントする' do
    list = FizzBuzz.generate_list
    assert_equal 1, list.first
    assert_equal 'Fizz', list[2]
    assert_equal 'Buzz', list[4]
    assert_equal 'FizzBuzz', list[14]
    assert_equal 'Buzz', list.last
  end

  test '3で割り切れる場合はFizzを返す' do
    assert_equal 'Fizz', FizzBuzz.generate(3)
    assert_equal 1, FizzBuzz.generate(1)
  end

  test '5で割り切れる場合はBuzzを返す' do
    assert_equal 'Buzz', FizzBuzz.generate(5)
  end

  test '3と5で割り切れる場合はFizzBuzzを返す' do
    assert_equal 'FizzBuzz', FizzBuzz.generate(15)
  end

  test "数字以外の場合は例外を発生させる" do
    assert_raise(RuntimeError)  {FizzBuzz.generate('a')}
  end
end

class FizzBuzz
  MAX_RANGE = 100
  FIZZ_BUZZ = 'FizzBuzz'
  FIZZ = 'Fizz'
  BUZZ = 'Buzz'

  def self.print
    begin
      puts generate_list
    rescue RuntimeError => error
      p error.backtrace
    end
  end

  def self.generate_list
    (1..MAX_RANGE).map(&method(:generate))
  end

  def self.generate(number)
    raise "数字以外は処理できません:#{number}" unless number.is_a? Numeric

    return FIZZ_BUZZ if (fizz?(number)).zero? && (buzz?(number)).zero?
    return FIZZ if (fizz?(number)).zero?
    return BUZZ if (buzz?(number)).zero?
    number
  end

  def self.fizz?(number)
    number % 3
  end

  def self.buzz?(number)
    number % 5
  end

  private_class_method :buzz?,:fizz?
end
